### ISSUE TITLE
Keep the selected block in a selected state when switching the sidebar to Document

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#13309]
When a block is selected, it stays in the selected state also when operating on other parts of the user interface. It is possible to click on any part of the sidebar or top bar, and the block is still selected.

For example, it's possible to activate menus in the top bar, and the block is still selected:

<img width="1279" alt="screenshot 2019-01-13 at 14 22 56" src="https://user-images.githubusercontent.com/1682452/51085894-9fa0ba00-173f-11e9-8952-dab4644918b4.png">

This is actually a good thing: the interaction with the block is consistent and predictable. There are no assumption in the UI behavior and the block selected state is in the users full control.

There's one relevant case where this is no longer true: when clicking the sidebar "Document" tab, the selected block loses its selected state.

To better clarify, I've made a GIF:

![block selected state](https://user-images.githubusercontent.com/1682452/51085944-60269d80-1740-11e9-95bf-94ca897d4eab.gif)

For mouse users:
it's maybe a minor annoyance but worth considering they have to move their mouse to the block, click into it to make it "selected" again, and move their mouse back to the sidebar again to be able to operate in the "block inspector" settings.

For keyboard users (including screen reader users, switch control users, etc.):
it's a serious breakage of keyboard navigation. Just imagine this flow:
- a block is selected
- users jump to the sidebar, either by using the `Ctrl + backtick` shortcut or using the navigation tools their assistive technology provides
- users switch the Sidebar to "Document" to explore its content or for any other reason
- they then switch the Sidebar to "Block"
- at that point, the block is not selected any longer and the inspector is empty
- to be able to operate again on the block settings, users have to tab multiple times through the whole editor interface to go back to the block and select it, which is a challenging task especially when there are many blocks
- then, they can finally jump again to the sidebar

As [reported by the accessibility team](https://make.wordpress.org/accessibility/2018/10/29/report-on-the-accessibility-status-of-gutenberg/), keyboard navigation between the content area and the sidebar is one of the main barriers for users with accessibility needs. Gutenberg should strive to make navigation through content as easy as possible.

Personally, I'm not sure there's even a good reason why the block should _not_ be selected when the sidebar switches to "Document". I'd tend to think the block selected state should be completely independent from the Sidebar state. 

Ideally, the block selected state should always be in full control of the user. The UI behavior shouldn't assume users intent.

I'd like to propose to keep the block selected when switching the sidebar to "Document". I've tried to have a look at how it works, and turned out it's a bit complicated as the sidebar switching logic is actually based on the presence of a selected block. I'd greatly appreciate if anyone with better knowledge of this implementation could have a look at this, as it would be a pretty good accessibility improvement.

**Comment by youknowriad** [comment]
There has been a lot of back and forth here. 

 - The proposal here has been tried on the plugin phase and we had feedback that the inspector was not discoverable enough.

 - We had an intermediary step where selecting the block, opens the inspector but selecting the document don't unselect the block. It was confusing for users.

I think if we need to change this, we should not be using tabs maybe? but anyway, this needs some design input.

**Comment by afercia** [comment]
To clarify: I haven't proposed to remove the switching mechanism. That should stay.

However, the block should always stay selected. Unselecting it creates a big accessibility barrier.

> It was confusing for users.

I'm guessing the testers group didn't include users with accessibility needs. I'd like to see this community start considering design and accessibility not as two separate things but as two fundamental parts of the same process.

**Comment by afercia** [comment]
Alternatively:

would it be possible to store a reference to the "last-selected" block and, when clicking again on the Block tab, restore its selected state?

**Comment by youknowriad** [comment]
Thanks for the clarification :). Let's get some feedback and proceed with the PR.

**Comment by jasmussen** [comment]
The sidebar header area has been through a number of iterations as we explored and responded to feedback. Back in the day, the sidebar header used breadcrumbs: https://github.com/WordPress/gutenberg/issues/1429

If you'll permit me to steal the screenshot from that thread:

![breadcrumbs](https://user-images.githubusercontent.com/1204802/51894505-3314f480-23a8-11e9-97c6-6b516b104ce8.png)

The behavior of ☝was _technically_ identical to what we have today, but it was just visually more descriptive of what was going on. Essentially the idea is that the sidebar is 100% contextual to what is selected. If nothing is selected, you inspect the document, if a block is selected, you inspect the block. 

Having used the tabs for a long time now, it feels like there are some benefits to that older interface, especially in light of recent developments with child blocks, as well as developments that are likely to be explored during phase 2, notably:

- The breadcrumb interface can show the hierarchy of the post/page
- The breadcrumb trail could be made parent-block aware, helping address #9628. I.e. `Post → Columns → Paragraph`. 
- In exploring enhancements to customization, the breadcrumb trail might even include the `Site` itself as a breadcrumb, so that the inspector can customize that

I understand the goal of this ticket, but because of the above, it seems like the most prudent first step would be a smaller step. The following sounds pragmatic, for example:

> would it be possible to store a reference to the "last-selected" block and, when clicking again on the Block tab, restore its selected state?

**Comment by draganescu** [comment]
Commenting here just to second @kjellr, found this ticket as I was about to open another one especially because on mobile is appears like a big bug, not a logical feature for the full screen sidebar :)

**Comment by afercia** [comment]
> is also very confusing on mobile

Yep, for many users with accessibility needs, navigating content is a linearized experience much like on mobile. In fact, it often happens that some designs fail for accessibility and mobile for the same reasons 🙂

**Comment by afercia** [comment]
> especially because on mobile is appears like a big bug

Yep, testing on a real mobile device highlights how much the current behavior is annoying 🙂On mobile, the Sidebar is full screen so it's pretty common to explore its content and click on "Document". Then, when clicking back to "Block" the sidebar is just empty.

See the screenshot below and imagine the following user flow:
- user is editing an image details (screenshot on the left)
- clicks on "Document" then on "Block"
- the sidebar is now empty (screenshot on the right)
- to continue editing the image details, user now has to:
  - close the sidebar
  - select the block again
  - click on Block Settings again

![Screenshot_20190427-151113](https://user-images.githubusercontent.com/1682452/56866633-c8417c80-69db-11e9-953a-ac4c7c39f520.jpg)

@gziolo (pinging you as you've asked for some accessibility actionable items) would be great to help this issue move on, when you have a chance. Thanks!

**Comment by gziolo** [comment]
> (pinging you as you've asked for some accessibility actionable items) would be great to help this issue move on, when you have a chance. Thanks!

Thanks, @afercia. I appreciate your ping. I added it my personal backlog to ensure I have it tracked. I might not have time to work in it in the upcoming 2-3 weeks, but I can assist anyone who volunteers to work on it :)

**Comment by gziolo** [comment]
We paired today with @draganescu to explore the proposed solution. There is one use case we wanted to confirm before we finalize the proposal. Should all types of selections be restored when switching from Document to Block settings? In particular:
- multiple blocks selected
- single block selected with text selection in RichText component

My feeling is that it would be great to cover all those use cases but I'm fine with whatever comes out from the discussion.

**Comment by kjellr** [comment]
>  Should all types of selections be restored when switching from Document to Block settings?

Yeah, I think all types of selections should be preserved if possible. That way the behavior is consistent all the time.

**Comment by draganescu** [comment]
This now has an open PR that fixes it, but I am not sure if all cases are covered. We'll see after it gets a review :)

**Comment by afercia** [comment]
@draganescu @gziolo thanks for working on this and for merging #15583 🙂It's a step in the right direction. I see two problems though.

1
- click on a block to select it
- the sidebar correctly switches to "Block"
- click the Block "tab" in the Sidebar header
- the Sidebar switches to "Document"
- expected: the Sidebar to stay on "Block"

2
- test using the keyboard
- navigate to a block to select it
- use the `Ctrl + backtick` or the alternative `Control + Option + N` shortcuts to navigate through the main regions and get to the Sidebar
- press the Tab key to focus the "Document" tab and press Enter
- the Sidebar correctly switches to "Document", focus stays on the "Document" tab
- press the Tab key to focus the "Block" tab and press Enter
- the Sidebar correctly switches to "Block"
- however, focus is moved to the previously selected block
- expected: focus should stay on the Sidebar "Block" tab

I do realize that probably the term "selected state" of the block can be source of confusion but the main point of this issue is that yes: the block should return to its original "selected state" but the focus should stay on the Sidebar tab.

The fact focus gets moved away would be completely unexpected, as users are operating on the Sidebar tabs and focus should stay there.

**Comment by afercia** [comment]
Minor: in the console, I see:
"Warning: React does not recognize the `wipeSelectedBlock` prop on a DOM element."

**Comment by gziolo** [comment]
I could reproduce all issues, thanks @afercia for raising them 😅 
Let's try to fix them by Friday all revert PR otherwise.

**Comment by draganescu** [comment]
Hi @afercia great finds, I was laser focused on "restoring selection" and didn't account for the focus too. I will redo this PR with a different approach suggested by @aduth and also keep in mind your points here!

**Comment by draganescu** [comment]
@afercia #18621 removed the clearing of selection on siderbar switch. So instead of preserving selection, we simply don't clear it anymore. This I believe solves this issue, while also not introducing anymore complexity.
